File: python/mirage/commands/builtin/generic/awk.py
import re
from collections.abc import AsyncIterator, Awaitable, Callable

from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.awk_types import (CMP_OP_PATTERN,
                                                       FIELD_PREFIX,
                                                       PRINT_STMT, AwkBlock,
                                                       AwkBoolOp, AwkBuiltin,
                                                       AwkCmpOp)
from mirage.commands.builtin.utils.stream import _resolve_source
from mirage.io.async_line_iterator import AsyncLineIterator
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


def _parse_program(program: str) -> tuple[str, str]:
    program = program.strip()
    if program.startswith("{"):
        return "", program[1:].rstrip("}")
    if "{" in program:
        idx = program.index("{")
        condition = program[:idx].strip()
        action = program[idx + 1:].rstrip("}").strip()
        return condition, action
    return program, ""


def _resolve_token(tok: str, field_map: dict[str, str]) -> str:
    if tok.startswith(FIELD_PREFIX):
        inner = tok[1:]
        if inner in field_map:
            ref = field_map[inner]
            return field_map.get(f"{FIELD_PREFIX}{ref}", "")
        return field_map.get(tok, tok)
    return field_map.get(tok, tok)


def _eval_simple(expr: str, field_map: dict[str, str]) -> bool:
    expr = expr.strip()
    m = re.match(rf"(.+?)\s*({CMP_OP_PATTERN})\s*(.+)", expr)
    if not m:
        if expr.startswith("/") and expr.endswith("/"):
            regex = expr[1:-1]
            return bool(re.search(regex, field_map.get(AwkBuiltin.REC, "")))
        val = _resolve_token(expr, field_map)
        try:
            return float(val) != 0
        except ValueError:
            return bool(val)
    lhs_raw, op, rhs_raw = m.group(1).strip(), m.group(2), m.group(3).strip()
    rhs_raw = rhs_raw.strip('"')
    lhs = _resolve_token(lhs_raw, field_map)
    rhs = _resolve_token(rhs_raw, field_map) if rhs_raw.startswith(
        FIELD_PREFIX) or rhs_raw in field_map else rhs_raw
    try:
        lhs_n, rhs_n = float(lhs), float(rhs)
        return {
            AwkCmpOp.EQ: lhs_n == rhs_n,
            AwkCmpOp.NE: lhs_n != rhs_n,
            AwkCmpOp.GT: lhs_n > rhs_n,
            AwkCmpOp.LT: lhs_n < rhs_n,
            AwkCmpOp.GE: lhs_n >= rhs_n,
            AwkCmpOp.LE: lhs_n <= rhs_n,
        }[AwkCmpOp(op)]
    except ValueError:
        if op == AwkCmpOp.EQ:
            return lhs == rhs
        if op == AwkCmpOp.NE:
            return lhs != rhs
        return False


def _eval_condition(condition: str, field_map: dict[str, str]) -> bool:
    condition = condition.strip()
    if condition == AwkBlock.BEGIN or condition == AwkBlock.END:
        return False
    if AwkBoolOp.OR in condition:
        return any(
            _eval_condition(p, field_map)
            for p in condition.split(AwkBoolOp.OR))
    if AwkBoolOp.AND in condition:
        return all(
            _eval_condition(p, field_map)
            for p in condition.split(AwkBoolOp.AND))
    return _eval_simple(condition, field_map)


def _eval_action(action: str, field_map: dict[str, str], fs: str) -> str:
    parts: list[str] = []
    for stmt in action.split(";"):
        stmt = stmt.strip()
        if not stmt:
            continue
        if stmt.startswith(PRINT_STMT):
            args = stmt[len(PRINT_STMT):].strip()
            if not args:
                parts.append(field_map.get(AwkBuiltin.REC, ""))
            else:
                tokens = re.split(r",\s*", args)
                vals: list[str] = []
                for tok in tokens:
                    tok = tok.strip()
                    if tok.startswith('"') and tok.endswith('"'):
                        vals.append(tok[1:-1])
                    else:
                        vals.append(_resolve_token(tok, field_map))
                parts.append(" ".join(vals))
    return "\n".join(parts) if parts else ""


def _build_field_map(line: str, fs: str, nr: int,
                     variables: dict[str, str]) -> dict[str, str]:
    fields = re.split(re.escape(fs) if len(fs) == 1 else fs,
                      line) if fs else line.split()
    field_map = {
        AwkBuiltin.REC: line,
        AwkBuiltin.NR: str(nr),
        AwkBuiltin.NF: str(len(fields)),
    }
    for i, f in enumerate(fields, 1):
        field_map[f"{FIELD_PREFIX}{i}"] = f
    for k, v in variables.items():
        field_map[k] = v
    return field_map


def _awk_eval_line(
    line: str,
    program: str,
    fs: str,
    variables: dict[str, str],
    nr: int,
) -> str | None:
    field_map = _build_field_map(line, fs, nr, variables)
    condition, action = _parse_program(program)
    if condition and not _eval_condition(condition, field_map):
        return None
    if not action:
        return line
    return _eval_action(action, field_map, fs)


def _eval_end_print(action: str, accum: dict, end_map: dict[str, str]) -> str:
    parts = []
    for stmt in action.split(";"):
        stmt = stmt.strip()
        if not stmt.startswith(PRINT_STMT):
            continue
        args = stmt[len(PRINT_STMT):].strip()
        if not args:
            continue
        tokens = re.split(r",\s*", args)
        vals = []
        for tok in tokens:
            tok = tok.strip()
            if tok.startswith('"') and tok.endswith('"'):
                vals.append(tok[1:-1])
            elif tok in accum:
                v = accum[tok]
                vals.append(str(int(v)) if v == int(v) else str(v))
            elif tok in end_map:
                vals.append(end_map[tok])
            else:
                vals.append(tok)
        parts.append(" ".join(vals))
    return "\n".join(parts)


def _eval_begin(action: str) -> str:
    parts = []
    for stmt in action.split(";"):
        stmt = stmt.strip()
        if not stmt.startswith(PRINT_STMT):
            continue
        args = stmt[len(PRINT_STMT):].strip()
        if not args:
            parts.append("")
            continue
        tokens = re.split(r",\s*", args)
        vals = []
        for tok in tokens:
            tok = tok.strip()
            if tok.startswith('"') and tok.endswith('"'):
                vals.append(tok[1:-1])
            else:
                vals.append(tok)
        parts.append(" ".join(vals))
    return "\n".join(parts)


def _parse_blocks(program: str) -> tuple[str, str, str]:
    begin = ""
    end = ""
    main = program

    begin_match = re.match(rf"{AwkBlock.BEGIN}\s*\{{([^}}]*)\}}\s*(.*)",
                           program, re.DOTALL)
    if begin_match:
        begin = begin_match.group(1).strip()
        main = begin_match.group(2).strip()

    end_match = re.search(rf"{AwkBlock.END}\s*\{{([^}}]*)\}}\s*$", main)
    if end_match:
        end = end_match.group(1).strip()
        main = main[:end_match.start()].strip()

    return begin, main, end


def _eval_accumulator(action: str, field_map: dict, accum: dict) -> None:
    for stmt in action.split(";"):
        stmt = stmt.strip()
        m = re.match(r"(\w+)\s*\+=\s*(.+)", stmt)
        if m:
            var, expr = m.group(1), m.group(2).strip()
            val = field_map.get(expr, expr)
            try:
                accum[var] = accum.get(var, 0) + float(val)
            except ValueError:
                pass


async def _awk_stream(
    source: AsyncIterator[bytes],
    program: str,
    fs: str,
    variables: dict[str, str],
) -> AsyncIterator[bytes]:
    begin, main, end = _parse_blocks(program)
    accum: dict[str, float] = {}
    nr = 0

    if begin:
        result = _eval_begin(begin)
        if result:
            yield (result + "\n").encode()

    async for line_bytes in AsyncLineIterator(source):
        nr += 1
        line = line_bytes.decode(errors="replace")
        if main:
            field_map = _build_field_map(line, fs, nr, variables)
            condition, action = _parse_program(main)
            if condition and not _eval_condition(condition, field_map):
                continue

            _eval_accumulator(action, field_map, accum)

            result = _awk_eval_line(line, main, fs, variables, nr)
            if result is not None and result:
                yield (result + "\n").encode()

    if end:
        end_map = {AwkBuiltin.NR: str(nr), AwkBuiltin.NF: "0"}
        result = _eval_end_print(end, accum, end_map)
        if result:
            yield (result + "\n").encode()


def _strip_mount(virtual_path: str, prefix: str) -> str:
    if prefix and virtual_path.startswith(prefix + "/"):
        return "/" + virtual_path[len(prefix):].lstrip("/")
    return virtual_path


async def awk(
    paths: list[PathSpec],
    texts: tuple[str, ...],
    *,
    read_bytes: Callable[..., Awaitable[bytes]],
    read_stream: Callable[..., AsyncIterator[bytes]],
    accessor: object = None,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    field_separator: str | None = None,
    variable_assignment: str | None = None,
    program_file: PathSpec | None = None,
    index: IndexCacheStore | None = None,
) -> tuple[ByteSource | None, IOResult]:
    if program_file is not None:
        f_path = program_file.strip_prefix
        program = (await read_bytes(accessor,
                                    f_path)).decode(errors="replace").strip()
        mount_prefix = paths[0].prefix if paths else program_file.prefix
        data_paths = [_strip_mount(t, mount_prefix)
                      for t in texts] + [p.strip_prefix for p in paths]
    elif texts:
        program = texts[0]
        data_paths = [p.strip_prefix for p in paths]
    else:
        raise ValueError(
            "awk: usage: awk [-F fs] [-v var=val] 'program' [file ...]")

    fs = field_separator if field_separator else " "
    variables: dict[str, str] = {}
    if variable_assignment and "=" in variable_assignment:
        key, val = variable_assignment.split("=", 1)
        variables[key] = val

    cache: list[str] = []
    if data_paths:
        source: AsyncIterator[bytes] = read_stream(accessor, data_paths[0])
        cache = [data_paths[0]]
    else:
        source = _resolve_source(stdin, "awk: missing input")

    return _awk_stream(source, program, fs, variables), IOResult(cache=cache)


__all__ = ["awk"]

File: typescript/packages/core/src/commands/builtin/github/awk.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubAccessor } from '../../../accessor/github.ts'
import { read as githubRead } from '../../../core/github/read.ts'
import { resolveGlob } from '../../../core/github/glob.ts'
import { stream as githubStream } from '../../../core/github/read.ts'
import { AsyncLineIterator } from '../../../io/async_line_iterator.ts'
import { IOResult } from '../../../io/types.ts'
import { PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { resolveSource } from '../utils/stream.ts'
import { lstripSlash } from '../../../util/slash.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { fatal: false })

function escapeRegex(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

function splitFields(line: string, fs: string): string[] {
  if (fs === '') return line.split(/\s+/).filter((s) => s !== '')
  const re = fs.length === 1 ? new RegExp(escapeRegex(fs)) : new RegExp(fs)
  return line.split(re)
}

function parseProgram(program: string): [string, string] {
  const trimmed = program.trim()
  if (trimmed.startsWith('{')) return ['', trimmed.slice(1).replace(/\}$/, '')]
  if (trimmed.includes('{')) {
    const idx = trimmed.indexOf('{')
    const condition = trimmed.slice(0, idx).trim()
    const action = trimmed
      .slice(idx + 1)
      .replace(/\}$/, '')
      .trim()
    return [condition, action]
  }
  return ['', trimmed]
}

function evalCondition(condition: string, fieldMap: Record<string, string>): boolean {
  const cond = condition.trim()
  if (cond === 'BEGIN' || cond === 'END') return false
  const patterns = [
    /(\$\d+|NR|NF)\s*==\s*(.+)/,
    /(\$\d+|NR|NF)\s*!=\s*(.+)/,
    /(\$\d+|NR|NF)\s*>\s*(.+)/,
    /(\$\d+|NR|NF)\s*<\s*(.+)/,
    /(\$\d+|NR|NF)\s*>=\s*(.+)/,
    /(\$\d+|NR|NF)\s*<=\s*(.+)/,
  ]
  for (const pat of patterns) {
    const m = pat.exec(cond)
    if (m !== null) {
      const lhsKey = m[1] ?? ''
      const rhsRaw = (m[2] ?? '').trim().replace(/^"|"$/g, '')
      const lhs = fieldMap[lhsKey] ?? ''
      const opMatch = /(==|!=|>=|<=|>|<)/.exec(cond)
      const op = opMatch !== null ? opMatch[1] : ''
      const lhsNum = Number.parseFloat(lhs)
      const rhsNum = Number.parseFloat(rhsRaw)
      if (!Number.isNaN(lhsNum) && !Number.isNaN(rhsNum)) {
        if (op === '==') return lhsNum === rhsNum
        if (op === '!=') return lhsNum !== rhsNum
        if (op === '>') return lhsNum > rhsNum
        if (op === '<') return lhsNum < rhsNum
        if (op === '>=') return lhsNum >= rhsNum
        if (op === '<=') return lhsNum <= rhsNum
      }
      if (op === '==') return lhs === rhsRaw
      if (op === '!=') return lhs !== rhsRaw
      return false
    }
  }
  if (cond.startsWith('/') && cond.endsWith('/')) {
    return new RegExp(cond.slice(1, -1)).test(fieldMap.$0 ?? '')
  }
  return true
}

function evalAction(action: string, fieldMap: Record<string, string>, fs: string): string {
  const parts: string[] = []
  for (const rawStmt of action.split(';')) {
    const stmt = rawStmt.trim()
    if (stmt === '') continue
    if (stmt.startsWith('print')) {
      const args = stmt.slice(5).trim()
      if (args === '') {
        parts.push(fieldMap.$0 ?? '')
      } else {
        const tokens = args.split(/,\s*/)
        const vals: string[] = []
        for (const raw of tokens) {
          const tok = raw.trim().replace(/^"|"$/g, '')
          vals.push(fieldMap[tok] ?? tok)
        }
        parts.push(vals.join(' '))
      }
    }
  }
  void fs
  return parts.join('\n')
}

function awkEvalLine(
  line: string,
  program: string,
  fs: string,
  variables: Record<string, string>,
  nr: number,
): string | null {
  const fields = splitFields(line, fs)
  const fieldMap: Record<string, string> = {
    $0: line,
    NR: String(nr),
    NF: String(fields.length),
  }
  for (let i = 0; i < fields.length; i++) fieldMap[`$${String(i + 1)}`] = fields[i] ?? ''
  for (const [k, v] of Object.entries(variables)) fieldMap[k] = v
  const [condition, action] = parseProgram(program)
  if (condition !== '' && !evalCondition(condition, fieldMap)) return null
  if (action === '') return line
  return evalAction(action, fieldMap, fs)
}

async function* awkStream(
  source: AsyncIterable<Uint8Array>,
  program: string,
  fs: string,
  variables: Record<string, string>,
): AsyncIterable<Uint8Array> {
  let nr = 0
  const iter = new AsyncLineIterator(source)
  for await (const lineBytes of iter) {
    nr += 1
    const line = DEC.decode(lineBytes)
    const result = awkEvalLine(line, program, fs, variables, nr)
    if (result !== null) yield ENC.encode(result + '\n')
  }
}

function stripMount(virtualPath: string, prefix: string): string {
  if (prefix !== '' && virtualPath.startsWith(prefix + '/')) {
    return '/' + lstripSlash(virtualPath.slice(prefix.length))
  }
  return virtualPath
}

async function awkCommand(
  accessor: GitHubAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const resolved =
    paths.length > 0 ? await resolveGlob(accessor, paths, opts.index ?? undefined) : paths
  const mountPrefix = resolved[0]?.prefix ?? ''
  const fFlag = typeof opts.flags.f === 'string' ? opts.flags.f : null
  let program: string
  let dataPaths: string[]
  if (fFlag !== null) {
    const programSpec = new PathSpec({
      original: fFlag,
      directory: fFlag,
      resolved: false,
      prefix: mountPrefix,
    })
    try {
      const bytes = await githubRead(accessor, programSpec, opts.index ?? undefined)
      program = DEC.decode(bytes).trim()
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err)
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
    }
    dataPaths = [
      ...texts.map((t) => stripMount(t, mountPrefix)),
      ...resolved.map((p) => p.stripPrefix),
    ]
  } else if (texts.length > 0 && texts[0] !== undefined) {
    program = texts[0]
    dataPaths = resolved.map((p) => p.stripPrefix)
  } else {
    return [
      null,
      new IOResult({
        exitCode: 1,
        stderr: ENC.encode(`awk: usage: awk [-F fs] [-v var=val] 'program' [file ...]\n`),
      }),
    ]
  }
  const fs = typeof opts.flags.F === 'string' ? opts.flags.F : ' '
  const variables: Record<string, string> = {}
  if (typeof opts.flags.v === 'string' && opts.flags.v.includes('=')) {
    const [key, val] = opts.flags.v.split('=', 2)
    if (key !== undefined && val !== undefined) variables[key] = val
  }
  const cache: string[] = []
  let source: AsyncIterable<Uint8Array>
  if (dataPaths.length > 0) {
    const firstPath = dataPaths[0]
    if (firstPath === undefined) return [null, new IOResult()]
    const spec = new PathSpec({
      original: firstPath,
      directory: firstPath,
      resolved: false,
      prefix: mountPrefix,
    })
    source = githubStream(accessor, spec, opts.index ?? undefined)
    cache.push(firstPath)
  } else {
    try {
      source = resolveSource(opts.stdin, 'awk: missing input')
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err)
      return [null, new IOResult({ exitCode: 1, stderr: ENC.encode(`${msg}\n`) })]
    }
  }
  return [awkStream(source, program, fs, variables), new IOResult({ cache })]
}

export const GITHUB_AWK = command({
  name: 'awk',
  resource: ResourceName.GITHUB,
  spec: specOf('awk'),
  fn: awkCommand,
})
